src/coding-guidelines/values.rst

```diff
@@ -5,3 +5,185 @@
+      Reading uninitialized memory:
+
+      - creates undefined behavior for most types,
+      - may violate niche or discriminant validity,
+      - may create invalid pointer values, or
+      - may produce values that violate type invariants.
+      
+      The sole exception is that unions work like C unions: any union field may be read, even if it was never written.
+      The resulting bytes must, however, form a valid representation for the field’s type,
+      which is not guaranteed if the union contains arbitrary data.
+
+   .. non_compliant_example::
+      :id: non_compl_ex_Qb5GqYTP6db1
+      :status: draft
+
+      This noncompliant example creates a value of type ``u32`` from uninitialized memory via 
+      `assume_init <https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html#method.assume_init>`_:
+
+      .. code-block:: rust
+
+         use std::mem::MaybeUninit;
+
+         let x: u32 = unsafe { MaybeUninit::uninit().assume_init() }; // UB
+
+   .. compliant_example::
+      :id: compl_ex_Ke869nSXuShU
+      :status: draft
+
+      Types such as ``u8``, ``u16``, ``u32``, and ``i128`` allow all possible bit patterns.
+      Provided the memory is initialized, there is no undefined behavior.
+
+      .. code-block:: rust
+
+         union U {
+             n: u32,
+             bytes: [u8; 4],
+         }
+
+        let u = U { bytes: [0xFF, 0xEE, 0xDD, 0xCC] };
+        let n = unsafe { u.n };   // OK — all bit patterns valid for u32
+
+   .. compliant_example::
+      :id: compl_ex_Ke869nSXuShV
+      :status: draft
+
+      This compliant example calles the ``write`` function to fully initialize low-level memory.
+
+      .. code-block:: rust
+
+         use std::mem::MaybeUninit;
+
+         let mut x = MaybeUninit::<u64>::uninit();
+         x.write(42);
+         let val = unsafe { x.assume_init() }; // OK — value was fully initialized
+
+   .. non_compliant_example::
+      :id: non_compl_ex_Qb5GqYTP6db2
+      :status: draft
+
+      Creating a reference from arbitrary or uninitialized bytes is always undefined behavior.
+      References must be valid, aligned, properly dereferenceable, and non-null.
+      Uninitialized memory cannot satisfy these invariants.
+
+      .. code-block:: rust
+
+         use std::mem::MaybeUninit;
+
+         let r: &u32 = unsafe { MaybeUninit::uninit().assume_init() }; // UB — invalid reference
+
+   .. non_compliant_example::
+      :id: non_compl_ex_Qb5GqYTP6db4
+      :status: draft
+
+      Not all bit patterns are valid pointers for all operations (e.g., provenance rules).
+      You cannot create a pointer from unspecified bytes.
+      Even a raw pointer type (e.g., ``*const T``) has validity rules.
+
+      .. code-block:: rust
+
+         use std::mem::MaybeUninit;
+
+         let p: *const u32 = unsafe { MaybeUninit::uninit().assume_init() }; // UB
```

Comment:
Yes but not for the reasons you said.

To explain:
```rust
let bytes: [u8; 8] = core::from_fn(|_| random());
// this transmute is fine, actually
let ptr: *mut Type = unsafe { mem::transmute(bytes) };
// UB only triggers here:
let value = unsafe { *ptr };
```

`ptr` has no provenance. Not just because it was obtained from random bytes instead, but because `transmute` of int-to-ptr does not preserve provenance on the integer. So this is a no-provenance pointer. It is permitted to create this, but the read requires a provenance to be attached to be correct, so it's not.